ddtrace/internal/processor/stats.py

```diff
@@ -0,0 +1,259 @@
+                serialized_bucket = {
+                    u"Name": six.ensure_text(name),
+                    u"Resource": six.ensure_text(resource),
+                    u"Synthetics": synthetics,
+                    u"HTTPStatusCode": http_status,
+                    u"Hits": stat_aggr.hits,
+                    u"TopLevelHits": stat_aggr.top_level_hits,
+                    u"Duration": stat_aggr.duration,
+                    u"Errors": stat_aggr.errors,
+                    u"OkSummary": DDSketchProto.to_proto(stat_aggr.ok_distribution).SerializeToString(),
+                    u"ErrorSummary": DDSketchProto.to_proto(stat_aggr.err_distribution).SerializeToString(),
+                }
+                if service:
+                    serialized_bucket[u"Service"] = six.ensure_text(service)
+                if _type:
+                    serialized_bucket[u"Type"] = six.ensure_text(_type)
+                bucket_aggr_stats.append(serialized_bucket)
+            serialized_buckets.append(
+                {
+                    u"Start": bucket_time_ns,
+                    u"Duration": self._bucket_size_ns,
+                    u"Stats": bucket_aggr_stats,
+                }
+            )
+
+        # Clear out buckets that have been serialized
+        for key in serialized_bucket_keys:
+            del self._buckets[key]
+
+        return serialized_buckets
+
+    def _flush_stats(self, payload):
+        # type: (bytes) -> None
+        try:
+            conn = get_connection(self._agent_url, self._timeout)
+            conn.request("PUT", self._endpoint, payload, self._headers)
+            resp = get_connection_response(conn)
+        except Exception:
+            log.error("failed to submit span stats to the Datadog agent at %s", self._agent_endpoint, exc_info=True)
+            raise
+        else:
+            if resp.status == 404:
+                log.error(
+                    "Datadog agent does not support tracer stats computation, disabling, please upgrade your agent"
+                )
+                self._enabled = False
+                return
+            elif resp.status >= 400:
+                log.error(
+                    "failed to send stats payload, %s (%s) (%s) response from Datadog agent at %s",
+                    resp.status,
+                    resp.reason,
+                    resp.read(),
+                    self._agent_endpoint,
+                )
+            else:
+                log.info("sent %s to %s", _human_size(len(payload)), self._agent_endpoint)
+
+    def periodic(self):
+        # type: (...) -> None
+
+        with self._lock:
+            serialized_stats = self._serialize_buckets()
+
+        if not serialized_stats:
+            # No stats to report, short-circuit.
+            return
+        raw_payload = {u"Stats": serialized_stats}  # type: Dict[str, Union[List[Dict], str]]
+        hostname = get_hostname()
```

Comment:
Can we get the hostname once, in the `__init__` instead? Or do we expect that it could potentially change at runtime?

Reply:
get_hostname() actually just caches the result anyway: https://github.com/DataDog/dd-trace-py/blob/0453062b8d0aff5a28bed7482375a7ba92c922b2/ddtrace/internal/hostname.py#L5-L13

Reply:
Ah right! There could still be some benefits in having a local reference, which would avoid a function call and a global lookup, but given that this is not on a critical path I'll leave it up to you.

Reply:
we could also save the cost of encoding it repeatedly, so i'll accept your suggestion!